src/libopensc/card-nqApplet.c

```diff
@@ -0,0 +1,556 @@
+};
+typedef struct nqapplet_driver_data * nqapplet_driver_data_ptr;
+
+static struct sc_card_driver nqapplet_driver = 
+{
+	"NQ-Applet",      // name
+	"nqapplet",      // short name
+	&nqapplet_operations,	// operations
+	NULL,           // atr table
+	0,              // nr of atr
+	NULL            // dll?
+};
+
+static const struct sc_card_error nqapplet_errors[] = 
+{
+	{ 0x6700, SC_ERROR_WRONG_LENGTH, "Invalid LC or LE"}, 
+	{ 0x6982, SC_ERROR_SECURITY_STATUS_NOT_SATISFIED, "Security status not satisfied" }, //TODO MK/DK??
+	{ 0x6985, SC_ERROR_NOT_ALLOWED,	"Invalid PIN or key" },
+	{ 0x6986, SC_ERROR_NOT_ALLOWED,	"Conditions of use not satisfied" },
+	{ 0x6A80, SC_ERROR_INVALID_ARGUMENTS, "Invalid parameters" },
+	{ 0x6A82, SC_ERROR_OBJECT_NOT_FOUND, "Data object not found"},
+	{ 0x6A84, SC_ERROR_NOT_ENOUGH_MEMORY, "Not enough memory" },
+	{ 0x6A86, SC_ERROR_INCORRECT_PARAMETERS, "Invalid P1 or P2" }, 
+	{ 0x6A88, SC_ERROR_INVALID_ARGUMENTS, "Wrong key ID" },
+	{ 0x6D00, SC_ERROR_FILE_NOT_FOUND, "Applet not found" }
+};
+
+/* convenience functions */
+
+static int init_driver_data(sc_card_t *card, u8 version_major, u8 version_minor, u8 * serial_nr, size_t cb_serial_nr)
+{
+	nqapplet_driver_data_ptr data = calloc(1, sizeof(struct nqapplet_driver_data));
+	if (data == NULL)
+	{
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+	}
+
+	data->version_major = version_major;
+	data->version_minor = version_minor;
+	data->key_reference = KEY_REFERENCE_NO_KEY;
+	memcpy(data->serial_nr, serial_nr, MIN(cb_serial_nr, sizeof(data->serial_nr)));
+	card->drv_data = (void*)data;
+	return SC_SUCCESS;
+}
+
+/**
+ * SELECT NQ-Applet, on success it returns the applet version and card serial nr.
+ *
+ * @param[in]     	card
+ * @param[out,opt]  version_major	Version major of the applet
+ * @param[out,opt]  version_minor	Version minor of the applet
+ * @param[out,opt]  serial_nr		Buffer to receive serial number octets
+ * @param[in]  		cb_serial_nr	Size of buffer in octets
+ * @param[out,opt]	serial_nr_len	The actual number of octet copied into serial_nr buffer
+ *
+ * @return SC_SUCCESS: The applet is present and selected.
+ * 
+ */
+static int select_nqapplet(sc_card_t *card, u8 *version_major, u8 *version_minor, u8 *serial_nr, size_t cb_serial_nr, size_t *serial_nr_len)
+{
+	int rv;
+	sc_context_t *ctx = card->ctx;
+	sc_apdu_t apdu;
+	u8 buffer[APPLET_VERSION_LEN+APPLET_MEMTYPE_LEN+APPLET_SERIALNR_LEN+2];
+	size_t cb_buffer = sizeof(buffer);
+	size_t cb_aid = sizeof(nqapplet_aid);
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_4_SHORT, 0xA4, 0x04, 0x00);
+	apdu.lc = cb_aid;
+	apdu.data = nqapplet_aid;
+	apdu.datalen = cb_aid;
+	apdu.resp = buffer;
+	apdu.resplen = cb_buffer;
+	apdu.le = cb_buffer;
+
+	rv = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(ctx, rv, "APDU transmit failure.");
+
+	rv = sc_check_sw(card, apdu.sw1, apdu.sw2);
+	LOG_TEST_RET(card->ctx, rv, "Card returned error");
+
+	if ( version_major != NULL )
```

Comment:
you should check first that the returned Lc was large enough otherwise you are reading uninitialized memory.

Reply:
Fixed it, added checking of apdu.resplen